src/functions/def.js

```diff
@@ -88,41 +88,65 @@ defineFunction({
     props: {
         numArgs: 0,
         allowedInText: true,
+        primitive: true,
     },
     handler({parser, funcName}) {
-        let arg = parser.gullet.consumeArgs(1)[0];
-        if (arg.length !== 1) {
-            throw new ParseError("\\gdef's first argument must be a macro name");
+        let tok = parser.gullet.popToken();
+        const name = tok.text;
+        if (/^(?:[\\{}$&#^_]|EOF)$/.test(name)) {
+            throw new ParseError("Expected a control sequence", tok);
         }
-        const name = arg[0].text;
-        // Count argument specifiers, and check they are in the order #1 #2 ...
+
         let numArgs = 0;
-        arg = parser.gullet.consumeArgs(1)[0];
-        while (arg.length === 1 && arg[0].text === "#") {
-            arg = parser.gullet.consumeArgs(1)[0];
-            if (arg.length !== 1) {
-                throw new ParseError(
-                    `Invalid argument number length "${arg.length}"`);
-            }
-            if (!(/^[1-9]$/.test(arg[0].text))) {
-                throw new ParseError(
-                    `Invalid argument number "${arg[0].text}"`);
-            }
-            numArgs++;
-            if (parseInt(arg[0].text) !== numArgs) {
-                throw new ParseError(
-                    `Argument number "${arg[0].text}" out of order`);
+        let insert;
+        const delimiters = [[]];
+        // <parameter text> contains no braces
+        while (parser.gullet.future().text !== "{") {
+            tok = parser.gullet.popToken();
+            if (tok.text === "#") {
+                // If the very last character of the <parameter text> is #, so that
+                // this # is immediately followed by {, TeX will behave as if the {
+                // had been inserted at the right end of both the parameter text
+                // and the replacement text.
+                if (parser.gullet.future().text === "{") {
+                    insert = parser.gullet.future();
+                    delimiters[numArgs].push("{");
+                    break;
+                }
+
+                // A parameter, the first appearance of # must be followed by 1,
+                // the next by 2, and so on; up to nine #’s are allowed
+                tok = parser.gullet.popToken();
+                if (!(/^[1-9]$/.test(tok.text))) {
+                    throw new ParseError(`Invalid argument number "${tok.text}"`);
+                }
+                if (parseInt(tok.text) !== numArgs + 1) {
+                    throw new ParseError(
+                        `Argument number "${tok.text}" out of order`);
+                }
+                numArgs++;
+                delimiters.push([]);
+            } else if (tok.text === "EOF") {
+                throw new ParseError("Expected a macro definition");
```

Comment:
How would this case arise?

Reply:
@kevinbarabash If the end is reached without the macro body, e.g., `\def\abc`.

Reply:
This seems really obvious now. 😅